virtool/uploads/data.py

```
@@ -0,0 +1,199 @@
+        uploads = []
+
+        async with AsyncSession(self._pg) as session:
+            if user:
+                filters.append(Upload.user == user)
+
+            if upload_type:
+                filters.append(Upload.type == upload_type)
+
+            if ready is not None:
+                filters.append(Upload.ready == ready)
+
+            results = await session.execute(select(Upload).filter(*filters))
+
+        for result in results.unique().scalars().all():
+            uploads.append(result.to_dict())
+
+        uploads = await apply_transforms(uploads, [AttachUserTransform(self._db)])
+
+        return [UploadMinimal(**upload) for upload in uploads]
+
+    async def create(
+        self,
+        name: str,
+        upload_type: str,
+        reserved: bool = False,
+        user: Optional[str] = None,
+    ) -> Upload:
+        """
+        Create an upload.
+        """
+
+        async with AsyncSession(self._pg) as session:
+            upload = Upload(
+                created_at=virtool.utils.timestamp(),
+                name=name,
+                ready=False,
+                removed=False,
+                reserved=reserved,
+                type=upload_type,
+                user=user,
+            )
+
+            session.add(upload)
+
+            await session.flush()
+
+            upload.name_on_disk = f"{upload.id}-{upload.name}"
+
+            upload = upload.to_dict()
+
+            await session.commit()
+
+        return Upload(**upload)
+
+    async def get(self, upload_id: int) -> Optional[Upload]:
+        """
+        Get a single upload by its ID.
+
+        :param upload_id: the upload's ID
+        :return: the upload
+        """
+        async with AsyncSession(self._pg) as session:
+            upload = (
+                await session.execute(
+                    select(Upload).filter_by(id=upload_id, removed=False)
+                )
+            ).scalar()
+
+            if not upload:
+                return None
```

Raise `ResourceNotFound` instead of returning none.